backends/gstreamer/src/audio_decoder.rs

```diff
@@ -194,49 +194,31 @@ impl AudioDecoder for GStreamerAudioDecoder {
-                                        return gst::FlowReturn::Error;
-                                    };
-
-                                    let caps = if let Some(caps) = sample.get_caps() {
-                                        caps
-                                    } else {
-                                        callbacks_.error(AudioDecoderError::Backend(
-                                            "Could not get caps from sample".to_owned(),
-                                        ));
-                                        return gst::FlowReturn::Error;
-                                    };
-
-                                    let audio_info = if let Some(audio_info) =
-                                        gst_audio::AudioInfo::from_caps(&caps)
-                                    {
-                                        audio_info
-                                    } else {
-                                        callbacks_.error(AudioDecoderError::Backend(
-                                            "AudioInfo failed".to_owned(),
-                                        ));
-                                        return gst::FlowReturn::Error;
-                                    };
+                                        gst::FlowError::Error
+                                    })?;
+
+                                    let audio_info = sample
+                                        .get_caps()
+                                        .and_then(|caps| {
+                                            gst_audio::AudioInfo::from_caps(caps.as_ref())
+                                        })
+                                        .ok_or_else(|| {
+                                            callbacks_.error(AudioDecoderError::Backend(
+                                                "Could not get caps from sample".to_owned(),
+                                            ));
+                                            gst::FlowError::Error
+                                        })?;
                                     assert_eq!(audio_info.channels(), 1);
-                                    let positions = if let Some(positions) = audio_info.positions()
-                                    {
-                                        positions
-                                    } else {
+                                    let positions = audio_info.positions().ok_or_else(|| {
                                         callbacks_.error(AudioDecoderError::Backend(
                                             "AudioInfo failed".to_owned(),
                                         ));
-                                        return gst::FlowReturn::Error;
-                                    };
+                                        gst::FlowError::Error
+                                    })?;
 
                                     for position in positions.iter() {
```

Comment:
Above you `assert_eq!(audio_info.channels(), 1);` and here you then iterate over the positions. There can only be one, or what am I missing here? :)